### PR TITLE
JN-938 Adding proxy question by answer mapping

### PR DESCRIPTION
**Pull request by pegahtah**
#### DESCRIPTION

This PR is for adding an answer mapping type for the proxy enrollment question and using the answer to that question in the preenroll survey to determine if the user is enrolling someone as a proxy or enrolling themselves
That kind of question is also added to the demo populate 

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

populate demo portal and start the regiesteration process. You should see a new question in the pre enroll asking you who are you enrolling? and if you selecting enrolling a child/dependent then your dashboard will be empty (for now) since tasks get assigned to your governed user and you are just a proxy

### CODE REVIEW

File: core/src/main/java/bio/terra/pearl/core/dao/survey/AnswerMappingDao.java [hunk 1]
```diff
@@ -25,4 +28,16 @@ public List<AnswerMapping> findBySurveyId(UUID surveyId) {
     public void deleteBySurveyId(UUID surveyId) {
         deleteByProperty("survey_id", surveyId);
     }
+    public Optional<AnswerMapping> findByTargetField(UUID surveyId, AnswerMappingTargetType targetType, String targetFieldName) {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/dao/survey/AnswerMappingDao.java:32]: nit: add blank line between methods
```suggestion

    public Optional<AnswerMapping> findByTargetField(UUID surveyId, AnswerMappingTargetType targetType, String targetFieldName) {
```

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java [hunk 2]
```diff
@@ -142,6 +152,47 @@ public HubResponse<Enrollee> enroll(EnvironmentName envName, String studyShortco
...
+        PreEnrollmentResponse preEnrollResponse = preEnrollmentResponseDao.find(preEnrollResponseId).get();
+        if (preEnrollResponse == null || !preEnrollResponse.isQualified()) {
+            return false;
+        }
+        UUID surveyId = preEnrollResponse.getSurveyId();
+        Optional<AnswerMapping> answerMappingForProxyOpt = answerMappingDao.findByTargetField(surveyId, AnswerMappingTargetType.PROXY, "isProxy");
+        if (answerMappingForProxyOpt.isEmpty()) {
+            return false;
+        }
+        String questionStableIdL = answerMappingForProxyOpt.get().getQuestionStableId();
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java]:
```suggestion
        String questionStableId = answerMappingForProxyOpt.get().getQuestionStableId();
```

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java [hunk 3]
```diff
@@ -142,6 +152,47 @@ public HubResponse<Enrollee> enroll(EnvironmentName envName, String studyShortco
...
+        try {
+            ObjectMapper objectMapper = new ObjectMapper();
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java]: we should inject the ObjectMapper via the constructor

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java [hunk 4]
```diff
@@ -142,6 +152,47 @@ public HubResponse<Enrollee> enroll(EnvironmentName envName, String studyShortco
...
+            try {
+                JsonNode rootNode = objectMapper.readTree(preEnrollResponse.getFullData());
+                if (rootNode.isArray()) {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java]: at least one, probably two, helper methods would make this code clearer.  e.g. "getAnswerByStableId(String fullDataString)"  -- these methods can be put in `SurveyParseUtils`

File: populate/src/main/resources/seed/portals/demo/studies/heartdemo/surveys/preEnroll.json [hunk 5]
```diff
@@ -80,6 +88,27 @@
             ]
           }
         ]
+      },
+      {
+        "elements": [
+          {
+            "type": "checkbox",
```
Comment (devonbush) [inline populate/src/main/resources/seed/portals/demo/studies/heartdemo/surveys/preEnroll.json]: this should be a radio button for now--we're not yet supporting enrolling both as subjects, right?  That will also simplify the parsing -- the value will just come in as the `booleanValue` on the question

**Review by devonbush — CHANGES_REQUESTED** [review]
Overall, looks great!  A couple of changes, mostly around pulling out helper methods.  And once you pull out the helper methods, we need to have tests around this.  Both unit tests for the helper, and a functional test for the full proxy enrollment in `EnrollmentServiceTests`, similar to testAnonymousPreEnroll

File: populate/src/main/resources/seed/portals/demo/studies/heartdemo/surveys/preEnroll.json (hunk 5 above)
Comment (pegahtah) [reply populate/src/main/resources/seed/portals/demo/studies/heartdemo/surveys/preEnroll.json]: Right now in Pepper both approaches exist. In CMI these are check boxes and if they choose both, they will first sign up as normal users and they can later add their governed user. But in AT it's as you described, radio buttons and only supporting one kind of enrollment. I will change it to radio buttons, it shouldn't be hard to change it later anyways

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java (hunk 2 above)
Comment (pegahtah) [reply core/src/main/java/bio/terra/pearl/core/service/workflow/EnrollmentService.java]: Oops typo! Thank you!

File: core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java [hunk 6]
```diff
@@ -144,4 +149,59 @@ public static String getUpstreamStableId(JsonNode calculatedValue) {
...
+            throw new RuntimeException(e);
+        }
+        return null;
+    }
+
+    /**
+     * This method is used to convert the answer to a survey to the returnClass. The method assumes the answer is a valid and is
+     * in the answerField of the node
+     * */
+    static <T> T convertQuestionAnswerToClass(JsonNode node, String answerField, Class<T> returnClass, ObjectMapper objectMapper) throws JsonProcessingException {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java]: nit: we generally include the visibility modifier -- I find it makes the code easier to read.
```suggestion
    protected static <T> T convertQuestionAnswerToClass(JsonNode node, String answerField, Class<T> returnClass, ObjectMapper objectMapper) throws JsonProcessingException {
```

File: core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java [hunk 7]
```diff
@@ -144,4 +149,59 @@ public static String getUpstreamStableId(JsonNode calculatedValue) {
...
+        // Attempt to use a constructor that takes a single String argument for other types
+        try {
+            Constructor<T> constructor = returnClass.getConstructor(String.class);
+            return constructor.newInstance(value);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("The provided returnClass does not have a String constructor that we can use.", e);
+        }
+    }
+
+    static String getQuestionStableId(JsonNode node) {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java]: same nit
```suggestion
    protected static String getQuestionStableId(JsonNode node) {
```

File: core/src/test/java/bio/terra/pearl/core/service/survey/SurveyServiceTests.java [hunk 8]
```diff
@@ -241,4 +251,75 @@ public void testProcessDerivedQuestions() {
...
+        String json3 = "{\"objectValue\": \"3\"}";
+        node = objectMapper.readTree(json3);
+        int integerResult = SurveyParseUtils.convertQuestionAnswerToClass(node, "objectValue", Integer.class, objectMapper);
+        assertEquals(3, integerResult);
+    }
+
+    @Test
+    void testGetSurveyAnswerFromPreEnrollSurveyJsonData() {
+        String jsonInput =
+                "[{\"createdAt\":1710527621.050828000,\"lastUpdatedAt\":1710527621.050828000,\"questionStableId\":"
```
Comment (devonbush) [inline core/src/test/java/bio/terra/pearl/core/service/survey/SurveyServiceTests.java]: please put this in a multiline String (""") for readability

**Review by devonbush — CHANGES_REQUESTED** [review]
Cool, code looks great minus a couple nits.  Please add a test in EnrollmentServiceTests that we can correctly pick up a preenrollment survey response and detect that it has indicated a porxy enrollment

File: core/src/testFixtures/java/bio/terra/pearl/core/factory/survey/PreEnrollmentSurveyFactory.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import bio.terra.pearl.core.model.survey.AnswerMapping;
+import bio.terra.pearl.core.model.survey.AnswerMappingMapType;
+import bio.terra.pearl.core.model.survey.AnswerMappingTargetType;
+import bio.terra.pearl.core.model.survey.PreEnrollmentResponse;
+import bio.terra.pearl.core.service.survey.PreEnrollmentResponseService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PreEnrollmentSurveyFactory {
```
Comment (devonbush) [inline core/src/testFixtures/java/bio/terra/pearl/core/factory/survey/PreEnrollmentSurveyFactory.java:16]: nice!  I'm glad you pulled these methods out into a factory

File: core/src/test/java/bio/terra/pearl/core/service/workflow/EnrollmentWorkflowTests.java [hunk 10]
```diff
@@ -249,6 +251,35 @@ public void testProxyEnrollingMultipleChild(TestInfo testInfo){
...
+        Assertions.assertTrue(enrollmentService.isProxyEnrollment(preEnrollmentResponseProxy.getId()));
+
+        String preEnrollmentSurveyResponseForRegularUser = """
+                [{"questionStableId": "proxyQuestion","surveyVersion":0,"viewedLanguage":"en","stringValue":"false"},
+                {"createdAt":1710527339.202811000,"lastUpdatedAt":1710527339.202811000,"questionStableId":"qualified","surveyVersion":0,"booleanValue":true}]""";
+        ParticipantUserFactory.ParticipantUserAndPortalUser userBundle = participantUserFactory.buildPersisted(portalEnv, getTestName(info));
+        PreEnrollmentResponse preEnrollmentResponse = preEnrollmentSurveyFactory.buildPersisted(getTestName(info), preEnrollmentSurvey.getId(), true, preEnrollmentSurveyResponseForRegularUser, userBundle.ppUser().getId(), userBundle.user().getId(), studyEnv.getId());
+        preEnrollmentResponseProxy.setQualified(true);
+        Assertions.assertFalse(enrollmentService.isProxyEnrollment(preEnrollmentResponse.getId()));
+
```
Comment (devonbush) [inline core/src/test/java/bio/terra/pearl/core/service/workflow/EnrollmentWorkflowTests.java:281]: this is great!  Now add one more check that confirms that we actually do a proxy enroll based on the results.

So, e.g.
```
  HubResponse<Enrollee> hubResponse = enrollmentService.enroll(studyEnv.getEnvironmentName(), studyShortcode,
                userProxyBundle.user(), userProxyBundle.ppUser(), preEnrollmentResponseProxy.getId());
// confirm that two enrollees were created, and only one is a subject  
assertThat(hubResponse.enrollee().isSubject(), equalTo(false));
assertThat(hubResponse.response().isSubject(), equalTo(true));

**Review by devonbush — APPROVED** [review]
cool, one small addition to the test, and then ship it!

File: core/src/test/java/bio/terra/pearl/core/service/workflow/EnrollmentWorkflowTests.java (hunk 10 above)
Comment (pegahtah) [reply core/src/test/java/bio/terra/pearl/core/service/workflow/EnrollmentWorkflowTests.java:281]: Done ✅

File: core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java [hunk 11]
```diff
@@ -144,4 +149,59 @@ public static String getUpstreamStableId(JsonNode calculatedValue) {
...
+     * This method is used to convert the answer to a survey to the returnClass. The method assumes the answer is a valid and is
+     * in the answerField of the node
+     * */
+    protected static <T> T convertQuestionAnswerToClass(JsonNode node, String answerField, Class<T> returnClass, ObjectMapper objectMapper) throws JsonProcessingException {
+        String objectValueString = node.get(answerField).asText();
+        // Direct conversion for String
+        if (returnClass == String.class) {
+            return returnClass.cast(objectValueString);
+        }
+        String value = objectMapper.readValue(objectValueString, String.class);
```
Comment (connorlbark) [inline core/src/main/java/bio/terra/pearl/core/service/survey/SurveyParseUtils.java:192]: Why not do `objectMapper.readValue(objectValueString, returnClass)`? I'm probably missing something, just curious

**Review by connorlbark — APPROVED** [review]
nice! works well. it looks like the java ci failed but it looks like a flaky test, maybe worth re-running just to check but otherwise 👍 !

**Pull request merged**